Add support eMIOS PWM pulse generation for S32K344

Dat-NguyenDuy opened the pull request:
This introduces PWM shim driver for NXP S32 eMIOS with APIs for pulse generation

Dat-NguyenDuy committed:
drivers: pwm: introduce PWM driver for NXP S32 EMIOS

This introduces PWM driver with supporting PWM output
APIs based on NXP S32 EMIOS peripheral. This supports
three mode: OPWFMB, OPWMCB and OPWMB.

OPWFMB uses internal counter, the new period and duty
cycle takes effect immediately.

OPWMCB and OPWMB use external counter as timebase, changing
PWM period at runtime will impact to all channels share the
same timebase. Also the new period and duty cycle take effect
in next period boundary of the timebase

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

Dat-NguyenDuy committed:
boards: mr_canhubk3: enable NXP S32 EMIOS PWM

In this board, there are eight PWM channels from
EMIOS 0 CH0 --> EMIOS 0 CH7 that can be used to
generate PWM pulse to outside of the board.

Moreover, there are three RGB leds that can use
PWM pins for blinking, faded leds

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

manuargue commented:
hi @Dat-NguyenDuy pls fix the failed compliance checks and builds

Dat-NguyenDuy commented:
> hi @Dat-NguyenDuy pls fix the failed compliance checks and builds

Hi, fixed the compliance

manuargue commented on dts/bindings/pwm/nxp,s32-emios-pwm.yaml:
```diff
@@ -0,0 +1,128 @@
+description: |
+  NXP S32 eMIOS PWM node for S32K3 SoCs. Each channel in eMIOS can be configured
```
does this binding only applies for S32K3? I'd expect it can be reused in other SoCs like S32Z

Dat-NguyenDuy replied:
Right now is fit for S32K3 but yes can be extended, updated

manuargue commented on modules/Kconfig.s32:
```diff
@@ -1,9 +1,13 @@
 	select HAS_CMSIS_CORE
 	depends on SOC_FAMILY_S32
+
+config NOCACHE_LOAD_MEMORY
```
perhaps we could consider making this available to other SoCs? similarly to `CONFIG_NOCACHE_MEMORY`

mmahadevan108 replied:
Can you add a description for this Kconfig

Dat-NguyenDuy replied:
When i revisited the bare-metal driver and had a discussion with the engineer from HAL team, he said that actually the variable is supposed and can be put in to a read only region. We will revisit the whole supporting features and if needed, will put a separate PR for this support

mmahadevan108 commented on dts/bindings/timer/nxp,s32-emios-timer.yaml:
```diff
@@ -0,0 +1,94 @@
+# Copyright 2023 NXP
```
Why is this file added. I don't see a timer driver. If this module is intended for a timer driver then this binding should be included as part of that PR.

Dat-NguyenDuy replied:
No, this module is not intended for timer driver (although it could), i will push the update to move it out of here

Dat-NguyenDuy replied:
I created separate misc driver for eMIOS that does initializes global configuration for eMIOS

danieldegrasse commented on drivers/pwm/pwm_nxp_s32_emios.c:
```diff
@@ -0,0 +1,445 @@
+#define EMIOS_PWM_MODE(mode)		DT_CAT(EMIOS_PWM_IP_, mode)
+#define EMIOS_PWM_MODE_WITH_FLAG(mode)	DT_CAT3(EMIOS_PWM_IP_, mode, _FLAG)
+#define EMIOS_PWM_POLARITY(mode)	DT_CAT(EMIOS_PWM_IP_, mode)
+#define EMIOS_PWM_PRESCALER_SRC(mode)	DT_CAT(EMIOS_PWM_IP_PS_SRC_, mode)
```
Nit- can we move these macros to the bottom of the file, where they are used?

Dat-NguyenDuy replied:
I cleaned up a little bit and move to bottom of the file

Dat-NguyenDuy committed:
drivers: misc: add NXP S32 eMIOS driver

This PR adds a misc driver for NXP S32 eMIOS peripheral.
eMIOS provides multiple unified channels (UCs), there are
several channels can be used as reference timebase
(master bus) for other channels. At this time, the
driver does initialize global configuration for eMIOS

Signed-off-by: Dat Nguyen Duy <dat.nguyenduy@nxp.com>

Dat-NguyenDuy commented on dts/bindings/timer/nxp,s32-emios-timer.yaml:
```diff
@@ -0,0 +1,94 @@
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  NXP S32 Enhanced Modular IO SubSystem (eMIOS) node for S32K3 SoCs.
```
Right now is fit for S32K3 but yes can be extended, updated

manuargue commented on drivers/misc/nxp_s32_emios/nxp_s32_emios.c:
```diff
@@ -0,0 +1,101 @@
+#define NXP_S32_EMIOS_GET_INSTANCE(n)								\
+	LISTIFY(__DEBRACKET eMIOS_INSTANCE_COUNT, NXP_S32_EMIOS_INSTANCE_CHECK, (|), n)
+
+/* Consider add comment for GlobalTimeBase */
```
nit: should this comment be removed?

Dat-NguyenDuy replied:
I removed

manuargue commented on drivers/misc/nxp_s32_emios/nxp_s32_emios.c:
```diff
@@ -0,0 +1,101 @@
+
+/* Consider add comment for GlobalTimeBase */
+#define NXP_S32_EMIOS_GENERATE_GLOBAL_CONFIG(n)							\
+	BUILD_ASSERT(DT_INST_PROP(n, clock_divider) <= 0xFFU,					\
```
Consider adding a macro definition for `0xff`

mmahadevan108 commented:
@Dat-NguyenDuy , please update `west.yml`

manuargue commented on dts/bindings/misc/nxp,s32-emios.yaml:
```diff
@@ -0,0 +1,95 @@
+    type: int
+    required: true
+    description: |
+      Clock divider value for the global prescaler. Could be in range [1 .. 256]
```
Isn't the build assertion in the shim driver checking that this value is between 0 and 255?

Dat-NguyenDuy replied:
Right, add check for range [1 256]

manuargue commented on dts/bindings/misc/nxp,s32-emios.yaml:
```diff
@@ -0,0 +1,95 @@
+      should be configured as required by application. For example, to enable bus A
+      of eMIOS instance 0:
+        master_bus {
+          emios0_bus_a: emios0_bus_a {
```
There are more properties required than the ones included in this snippet. Perhaps add them all to have a complete example.

Dat-NguyenDuy replied:
Yep, i added for more details

manuargue reviewed (commented):
Some minor nits left from my side

manuargue reviewed (dismissed):
All my comments addressed. Failed builds seems to be caused by other boards unrelated to these changes.

Dat-NguyenDuy commented:
For imxrt, samples\drivers\led_pwm is build failing look like because of this updates https://github.com/zephyrproject-rtos/hal_nxp/pull/259. The test still passes on main because that one was merged in hal_nx[ but somehow the manifest in main was not updated accordingly

Dat-NguyenDuy commented:
Rebase main to pick up https://github.com/zephyrproject-rtos/zephyr/pull/62050, that fixed the issue causes some nrf boards build failure

mmahadevan108 commented:
Could you please rebase. The CI failure related to `RT 1170` has been addressed and merged.

Dat-NguyenDuy commented:
> Could you please rebase. The CI failure related to `RT 1170` has been addressed and merged.

Thanks, i rebased

manuargue reviewed (approved):
All my comments address, approved.

@bperseghetti if you can spend some time reviewing these changes it would be valuable to get feedback from you, since you are working on the integration of this board in a larger system.

bperseghetti commented:
> All my comments address, approved.
> 
> @bperseghetti if you can spend some time reviewing these changes it would be valuable to get feedback from you, since you are working on the integration of this board in a larger system.

I think overall it looks good and is a much needed add for the board/SoC. I think my only things are naming nits here and there but those are just preferential, I don't see any issues from reviewing over it.
